# Importing the libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

# Importing the dataset

In [ ]:
df = pd.read_csv("/content/Housing.csv")

#Displaying the dataset

In [ ]:
df

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

In [ ]:
# Displaying the first 10 rows of the dataset
df.head(10)

price   area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000   7420         4          2        3      yes        no       no   
1  12250000   8960         4          4        4      yes        no       no   
2  12250000   9960         3          2        2      yes        no      yes   
3  12215000   7500         4          2        2      yes        no      yes   
4  11410000   7420         4          1        2      yes       yes      yes   
5  10850000   7500         3          3        1      yes        no      yes   
6  10150000   8580         4          3        4      yes        no       no   
7  10150000  16200         5          3        2      yes        no       no   
8   9870000   8100         4          1        2      yes       yes      yes   
9   9800000   5750         3          2        4      yes       yes       no   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished  
5              no             yes        2      yes   semi-furnished  
6              no             yes        2      yes   semi-furnished  
7              no              no        0       no      unfurnished  
8              no             yes        2      yes        furnished  
9              no             yes        1      yes      unfurnished

In [ ]:
# Displaying the last 15 rows of the dataset
df.tail(15)

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
530  2240000  1950         3          1        1       no        no       no   
531  2233000  5300         3          1        1       no        no       no   
532  2135000  3000         2          1        1       no        no       no   
533  2100000  2400         3          1        2      yes        no       no   
534  2100000  3000         4          1        2      yes        no       no   
535  2100000  3360         2          1        1      yes        no       no   
536  1960000  3420         5          1        2       no        no       no   
537  1890000  1700         3          1        2      yes        no       no   
538  1890000  3649         2          1        1      yes        no       no   
539  1855000  2990         2          1        1       no        no       no   
540  1820000  3000         2          1        1      yes        no      yes   
541  1767150  2400         3          1        1       no        no       no   
542  1750000  3620         2          1        1      yes        no       no   
543  1750000  2910         3          1        1       no        no       no   
544  1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
530             yes              no        0       no      unfurnished  
531              no             yes        0      yes      unfurnished  
532              no              no        0       no      unfurnished  
533              no              no        0       no      unfurnished  
534              no              no        0       no      unfurnished  
535              no              no        1       no      unfurnished  
536              no              no        0       no      unfurnished  
537              no              no        0       no      unfurnished  
538              no              no        0       no      unfurnished  
539              no              no        1       no      unfurnished  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished

In [ ]:
df.shape

(545, 13)

It is understood from this that there are 545 rows and 13 columns

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


It is understood from this info that there are no null values. So we need not manipulate the data to remove the empty values.

In [ ]:
df.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

In [ ]:
df.columns

Index(['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'furnishingstatus'],
      dtype='object')

We now have a non-null dataframe with no outliers. But the object-type categorical data needs to be converted to some numeric-valued data to understand and analyze the dataset and perform Linear Regresion.

In [ ]:
cat_var_list =  ['mainroad', 'guestroom', 'basement', 'hotwaterheating', 'airconditioning', 'prefarea']

def binary_map(x):
    return x.map({'yes': 1, "no": 0})

df[cat_var_list] = df[cat_var_list].apply(binary_map)

df

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  \
0    13300000  7420         4          2        3         1          0   
1    12250000  8960         4          4        4         1          0   
2    12250000  9960         3          2        2         1          0   
3    12215000  7500         4          2        2         1          0   
4    11410000  7420         4          1        2         1          1   
..        ...   ...       ...        ...      ...       ...        ...   
540   1820000  3000         2          1        1         1          0   
541   1767150  2400         3          1        1         0          0   
542   1750000  3620         2          1        1         1          0   
543   1750000  2910         3          1        1         0          0   
544   1750000  3850         3          1        2         1          0   

     basement  hotwaterheating  airconditioning  parking  prefarea  \
0           0                0                1        2         1   
1           0                0                1        3         0   
2           1                0                0        2         1   
3           1                0                1        3         1   
4           1                0                1        2         0   
..        ...              ...              ...      ...       ...   
540         1                0                0        2         0   
541         0                0                0        0         0   
542         0                0                0        0         0   
543         0                0                0        0         0   
544         0                0                0        0         0   

    furnishingstatus  
0          furnished  
1          furnished  
2     semi-furnished  
3          furnished  
4          furnished  
..               ...  
540      unfurnished  
541   semi-furnished  
542      unfurnished  
543        furnished  
544      unfurnished  

[545 rows x 13 columns]

The categorical data has been converted to numeric data. Next we need to convert 'furnishingstatus' values to some type of numeric values as well. In the above conversion, the columns were 2-level categorical variables so it was easy to convert with yes corresponding to 1 and no corresponding to 0. But furnishingstatus has 3 levels. When you have a categorical variable with n-levels, the idea is to create a dummy variable by building ‘n-1’ variables, indicating the levels. Pandas has get_dummies method for it.

In [ ]:
furnishing_status = pd.get_dummies(df['furnishingstatus'])

furnishing_status

furnished  semi-furnished  unfurnished
0            1               0            0
1            1               0            0
2            0               1            0
3            1               0            0
4            1               0            0
..         ...             ...          ...
540          0               0            1
541          0               1            0
542          0               0            1
543          1               0            0
544          0               0            1

[545 rows x 3 columns]

We don’t need three columns. We can drop the furnished column, as it can be identified with just the last two columns values where:
00 - furnished ,
01 - unfurnished ,
10 - semi-furnished

In [ ]:
# Dropping the first column from furnishing_status dataset
furnishing_status = pd.get_dummies(df['furnishingstatus'], drop_first = True)

# Adding the converted finishing status to the original df dataframe
df = pd.concat([df, furnishing_status], axis = 1)

# Dropping 'furnishingstatus' as we have created the dummies for it
df.drop(['furnishingstatus'], axis = 1, inplace = True)

df

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  \
0    13300000  7420         4          2        3         1          0   
1    12250000  8960         4          4        4         1          0   
2    12250000  9960         3          2        2         1          0   
3    12215000  7500         4          2        2         1          0   
4    11410000  7420         4          1        2         1          1   
..        ...   ...       ...        ...      ...       ...        ...   
540   1820000  3000         2          1        1         1          0   
541   1767150  2400         3          1        1         0          0   
542   1750000  3620         2          1        1         1          0   
543   1750000  2910         3          1        1         0          0   
544   1750000  3850         3          1        2         1          0   

     basement  hotwaterheating  airconditioning  parking  prefarea  \
0           0                0                1        2         1   
1           0                0                1        3         0   
2           1                0                0        2         1   
3           1                0                1        3         1   
4           1                0                1        2         0   
..        ...              ...              ...      ...       ...   
540         1                0                0        2         0   
541         0                0                0        0         0   
542         0                0                0        0         0   
543         0                0                0        0         0   
544         0                0                0        0         0   

     semi-furnished  unfurnished  
0                 0            0  
1                 0            0  
2                 1            0  
3                 0            0  
4                 0            0  
..              ...          ...  
540               0            1  
541               1            0  
542               0            1  
543               0            0  
544               0            1  

[545 rows x 14 columns]

In [ ]:
X = df[['area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'semi-furnished','unfurnished']]
y = df.price

X contains all the columns except price and y contains only the price column

In [ ]:
X

area  bedrooms  bathrooms  stories  mainroad  guestroom  basement  \
0    7420         4          2        3         1          0         0   
1    8960         4          4        4         1          0         0   
2    9960         3          2        2         1          0         1   
3    7500         4          2        2         1          0         1   
4    7420         4          1        2         1          1         1   
..    ...       ...        ...      ...       ...        ...       ...   
540  3000         2          1        1         1          0         1   
541  2400         3          1        1         0          0         0   
542  3620         2          1        1         1          0         0   
543  2910         3          1        1         0          0         0   
544  3850         3          1        2         1          0         0   

     hotwaterheating  airconditioning  parking  prefarea  semi-furnished  \
0                  0                1        2         1               0   
1                  0                1        3         0               0   
2                  0                0        2         1               1   
3                  0                1        3         1               0   
4                  0                1        2         0               0   
..               ...              ...      ...       ...             ...   
540                0                0        2         0               0   
541                0                0        0         0               1   
542                0                0        0         0               0   
543                0                0        0         0               0   
544                0                0        0         0               0   

     unfurnished  
0              0  
1              0  
2              0  
3              0  
4              0  
..           ...  
540            1  
541            0  
542            1  
543            0  
544            1  

[545 rows x 13 columns]

In [ ]:
y

0      13300000
1      12250000
2      12250000
3      12215000
4      11410000
         ...   
540     1820000
541     1767150
542     1750000
543     1750000
544     1750000
Name: price, Length: 545, dtype: int64

Splitting the data into training and testing datasets

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2)

Building the model

In [ ]:
model = LinearRegression()

In [ ]:
model.fit(X_train, y_train)

LinearRegression()

Predicting using testing dataset

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
y_pred

array([5726176.72139726, 2568343.57042146, 3265046.3173656 ,
       5725083.10259781, 4822639.40406644, 7139663.9342049 ,
       3454817.76703354, 6489288.40308572, 4902814.84470951,
       4751023.72808717, 3120179.00038544, 3964841.52628234,
       7098461.5153297 , 3338996.11326773, 2686249.31447646,
       2708858.11190909, 3603121.20743679, 3168240.13035836,
       7176786.77342935, 4897203.78333021, 6206832.89163363,
       5858413.23323434, 4743080.39424304, 2719499.87642961,
       3254614.05463154, 2576023.73702569, 2315622.37148344,
       5361958.58175084, 3858501.80781383, 4138520.34900347,
       6046367.18927117, 6605895.98488136, 3925083.0400962 ,
       5869719.74310651, 5188645.62556989, 3666518.07743132,
       3360674.86176634, 5151883.78902616, 2381924.10025771,
       6766672.16823677, 6342405.48464118, 4980852.16430351,
       3014637.14956488, 4822175.4443146 , 9871531.37328928,
       7874810.02861075, 3690392.6272782 , 3331970.52332242,
       4533693.90144082,

In [ ]:
y_test 

110    6090000
482    2940000
475    2961000
167    5285000
289    4200000
        ...   
330    3990000
297    4200000
247    4550000
138    5740000
181    5215000
Name: price, Length: 109, dtype: int64

In [ ]:
r2_score(y_test, y_pred)

0.6509275543092705

The accuracy of my model is 76%

In [ ]:
#Predicting the price with sample input
model.predict([[7420,4,2,3,1,0,0,0,1,2,1,0,0]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([8123047.54001181])

# *-------  Thank You  -------*